mimirtool: Improve mimirtool docs for migrate-utf8 (#8576)
grafana/mimir: docs/sources/mimir/manage/tools/mimirtool.md

mimirtool alertmanager delete
```

> #### Migrate Alertmanager configuration for UTF-8 in Mimir 2.12 and newer

Reviewer: ```suggestion
#### Migrate Alertmanager configuration for UTF-8 in Mimir 2.12 and later
```

Reviewer: We use "later" instead of "newer" or "higher" when talking about versions.

> This requires mimirtool version 2.12 or newer. You can check the version of mimirtool with
> `mimirtool version`.

Reviewer: ```suggestion
This configuration requires mimirtool version 2.12 or later. To check your version of mimirtool, run
`mimirtool version`.
```

In accordance with [prometheus/prometheus#13095](https://github.com/prometheus/prometheus/issues/13095)
and [prometheus/alertmanager#3486](https://github.com/prometheus/alertmanager/issues/3486), Mimir is

> adding support for UTF-8. To support UTF-8 in alerts, routes, silences and inhibition rules,

Reviewer: ```suggestion
adding support for UTF-8. To support UTF-8 in alerts, routes, silences, and inhibition rules,
```

Alertmanager has added a new parser for matchers that has a number of backwards incompatible changes.

> More information about these changes can be found [here](https://prometheus.io/docs/alerting/latest/configuration/#label-matchers).

Reviewer: ```suggestion
For more information about these changes, refer to [Label matchers](https://prometheus.io/docs/alerting/latest/configuration/#label-matchers) in the Prometheus documentation.
```

Reviewer: We might want to remove the anchor and just link to https://prometheus.io/docs/alerting/latest/configuration/. If Prometheus updates this heading, our link could easily break.

Author: Ah. I think it's important to link to that specific section of the page. Is there some workaround we could do here?

Author: I just checked and we included `#label-matchers` in #8572.

Reviewer: We can include the anchor if it's really important. Just something to be aware of.

The migrate-utf8 command migrates an existing Alertmanager configuration in preparation for when
UTF-8 is enabled in a Mimir installation. To do this, it translates matchers that are incompatible

> with UTF-8 into equivalent matchers that are compatible. This translation is backwards compatible
> and does not change the behavior of existing matchers, and works even in Mimir installations that
> do not have UTF-8 enabled.

Reviewer: ```suggestion
with UTF-8 into equivalent matchers that are compatible. This translation is backwards compatible,  does not change the behavior of existing matchers, and works even in Mimir installations that
do not have UTF-8 enabled.
```

> A lot of Alertmanager configurations will not need migrating. You can verify if an existing
> Alertmanager configuration needs to be migrated for UTF-8 by running the verify command:

Reviewer: ```suggestion
You don't need to migrate all Alertmanager configurations. To verify if you need to migrate an existing
Alertmanager configuration for UTF-8, run the verify command:
```

Author: > You don't need to migrate all Alertmanager configurations.

Some users might, it depends on what's in their configuration. We found that the number of affected users is very small (single digit %), but if a user has an incompatible line in all their configurations, then all will need migrating. If that makes sense?

Author: What about:

> Whether an Alertmanager configuration needs to be migrated depends on whether it contains matchers
incompatible with UTF-8. You can verify if an existing Alertmanager configuration needs to be
migrated for UTF-8 by running the verify command:

Reviewer: I like this. Here's a small suggestion to avoid the double whether, as well as passive voice.

"The need to migrate an Alertmanager configuration depends on whether it contains matchers
incompatible with UTF-8. To determine whether you need to migrate an existing configuration, run the verify command:"

```bash
mimirtool alertmanager verify <config_file> [template_files...]
```

> If the command succeeds without any warnings, the configuration is compatible with UTF-8 and no
> migration is required.

Reviewer: ```suggestion
If the command succeeds without any warnings, the configuration is compatible with UTF-8 and you don't need to migrate it.
```

> However, if the command prints the following warning(s):
> 
> ```
> Alertmanager is moving to a new parser for labels and matchers, and this input is incompatible. Alertmanager has instead parsed the input using the classic matchers parser as a fallback. To make this input compatible with the UTF-8 matchers parser please make sure all regular expressions and values are double-quoted. If you are still seeing this message please open an issue.
> ```
> 
> then the Alertmanager configuration needs to be migrated with the migrate-utf8 command.

Reviewer: ```suggestion
However, if the command prints the following warning, then you need to migrate the Alertmanager configuration with the migrate-utf8 command.

```
Alertmanager is moving to a new parser for labels and matchers, and this input is incompatible. Alertmanager has instead parsed the input using the classic matchers parser as a fallback. To make this input compatible with the UTF-8 matchers parser, ensure that all regular expressions and values are double-quoted. If you still see this message, open an issue.
```

```

Reviewer: We try to avoid using "please" in documentation and error messages.

Reviewer: I think it reads a bit better if we show the error message below the conditional sentence.

Author: Ah. I can't remove the `please` as this error message comes from Prometheus. But I made all other changes.

When you use the `--output-dir` flag, the command only writes the output to files and doesn't print the configuration to the console.
{{< /admonition >}}

> Once you have migrated an Alertmanager configuration, verify it using the verify command:

Reviewer: ```suggestion
After you migrate an Alertmanager configuration, use the verify command to verify it:
```

```bash
mimirtool alertmanager verify <config_file> [template_files...]
```

> It should succeed without warnings.

Reviewer: I think we can probably remove this line- it's self-evident.

Author: I thought about this, but in this specific case I would like to make it as explicit as possible, as it's quite an important step for customers to run against their configurations.

Author: I thought about this, but in this specific case I would like to make it as explicit as possible, as it's quite an important step for customers to run against their configurations.

> You can also verify an Alertmanager configuration is compatible with UTF-8 by running the
> verify command with the `--utf8-strict-mode` flag:

Reviewer: ```suggestion
You can also run the
verify command with the `--utf8-strict-mode` flag to verify if an Alertmanager configuration is compatible with UTF-8 
```

```bash
mimirtool alertmanager verify <config_file> [template_files...] --utf8-strict-mode
```

> It will print a warning with the text `UTF-8 mode enabled` to let you know the command is using
> UTF-8 strict mode to validate the Alertmanager configuration.

Reviewer: ```suggestion
This command prints a warning with the text `UTF-8 mode enabled` to let you know the command is using
UTF-8 strict mode to validate the Alertmanager configuration.
```

Reviewer: We try to keep docs in present tense and avoid using "will".

> When using the `--utf8-strict-mode` flag, the verify command does not print any warnings if a

Reviewer: ```suggestion
When using the `--utf8-strict-mode` flag, the verify command doesn't print any warnings if a
```

> configuration is incompatible with UTF-8. Instead, the command will exit with an error. For example:

Reviewer: ```suggestion
configuration is incompatible with UTF-8. Instead, the command exits with an error. For example:
```
